pkg/merkle/proof.go

```diff
@@ -13,65 +12,43 @@ const (
 )
 
 var (
-	ErrDuplicateIndices   = errors.New("duplicate indices")
-	ErrIndicesOutOfBounds = errors.New("indices out of bounds")
-	ErrInvalidRange       = errors.New("invalid range")
-	ErrInvalidLeafCount   = errors.New("invalid leaf count")
-	ErrNilProof           = errors.New("nil proof")
-	ErrNilRoot            = errors.New("nil root")
-	ErrNilElement         = errors.New("nil element")
-	ErrNoElements         = errors.New("no elements")
-	ErrNoIndices          = errors.New("no indices")
-	ErrNoProofs           = errors.New("no proofs provided")
+	ErrInvalidLeafCount     = errors.New("invalid leaf count")
+	ErrNilProof             = errors.New("nil proof")
+	ErrNilRoot              = errors.New("nil root")
+	ErrNoProofs             = errors.New("no proofs provided")
+	ErrInvalidStartingIndex = errors.New("invalid starting index")
 )
 
-type MultiProof struct {
-	values    IndexedValues
-	proofs    []Node
-	leafCount int
-}
-
-func (p *MultiProof) GetLeafCount() int {
-	return p.leafCount
-}
+type ProofElement []byte
 
-func (p *MultiProof) GetProofs() []Node {
-	return p.proofs
-}
-
-func (p *MultiProof) GetValues() IndexedValues {
-	return p.values
+type MultiProof struct {
+	startingIndex int
+	leaves        []Leaf
+	proofElements []ProofElement
 }
 
-type IndexedValues []IndexedValue
-
-type IndexedValue struct {
-	value []byte
-	index int
+func (p *MultiProof) GetStartingIndex() int {
+	return p.startingIndex
 }
 
-func (iv IndexedValue) GetValue() []byte {
-	return iv.value
+func (p *MultiProof) GetLeafCount() int {
+	return Bytes32ToInt(p.proofElements[0])
 }
 
-func (iv IndexedValue) GetIndex() int {
-	return iv.index
+func (p *MultiProof) GetProofElements() []ProofElement {
+	return p.proofElements
 }
 
-func (iv IndexedValues) Indices() []int {
-	indices := make([]int, len(iv))
-	for i, v := range iv {
-		indices[i] = v.index
-	}
-	return indices
+func (p *MultiProof) GetLeaves() []Leaf {
+	return p.leaves
 }
 
-func (iv IndexedValues) ToLeaves() []Leaf {
-	leaves := make([]Leaf, len(iv))
-	for i, v := range iv {
-		leaves[i] = Leaf(v.value)
+func NewMerkleProof(startingIndex int, leaves []Leaf, proofElements []ProofElement) *MultiProof {
```

Comment:
This func is never used except in tests, meaning that the test could use the respective GenerateMultiProof function and this removed.

Reply:
> I thought other nodes (or even via an API), would want to take a startingIndex, leaves, and proofElements to build a MultiProof and check it, so this method allows that to be possible in a node without needing all the leaves to build a tree first

Up to you! Realistically this not going to happen, similarly to what was discussed around generating proofs with indices.